enhancements/authentication/allow-users-to-manage-their-own-tokens.md

```diff
@@ -0,0 +1,187 @@
+work with OpenShift services, their capabilities to see the tokens they
+created and to handle them is a bit lacking. This enhancement describes
+how the self-management of users' access tokens can be improved.
+
+## Motivation
+
+It is quite easy for users to retrieve tokens but it is also not so hard
+for them to be able to lose access to them - people don't usually search
+their cookies for the value of the token, or don't keep history of the
+tokens they retrieve by issuing the `oc login` command multiple times.
+
+### Goals
+
+Allow users to (using either console/CLI):
+- list their own tokens
+- view details of a specific token
+- delete their own tokens
+
+### Non-Goals
+
+- enable handling sessions of services that use the OpenShift access tokens
+
+## Proposal
+
+### User Stories [optional]
+
+#### Story 1
+A user is able to see all of the tokens that were issued for them, including
+the name of the OAuth2 client that requested it and the scopes of it, making
+it easier for them to audit their accesses.
+
+#### Story 2
+A user gets a friendlier way to remove their own tokens,
+`oc delete oauthaccesstoken <sha256-tokenhash>` uses generic deletion code
+which attempts to `list` the token after its deletion, and that ends up
+with a `forbidden` message since the user is usually not allowed to `list`
+their tokens.
+
+#### Story 3
+A user is able to perform actions from Story 1 and Story 2 in both the web
+console and the CLI clients.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The OAuth API server gets a new API endpoint - `useroauthaccesstokens`. This endpoint
+allows to `get`, `list`, `watch` and `delete` tokens that belong to the current user.
+The user is determined based on the context of the request, as supplied by the generic
+API server logic.
+
+Upon receiving a REST request to the new endpoint, the API server internally
+wires the get/list/watch/delete requests to the `oauthaccesstoken` API endpoint.
+It then converts the results and returns them as `useroauthaccesstoken` type.
+
+To list/watch the tokens internally, a field selector of the form `userName=<userName>`
+is used. The `userName` field selector must not be overridden by any
+user-specified selector.
+
+When getting or deleting a token, there must be a strict match of `token.userName`
+and the `userName` retrieved as described above. If such a match does not occur,
+the API server must return "404 - Not Found" so that it is not possible to guess
+other people's token object names.
+
+#### Roles modifications
+
+The `basic-user` clusterrole should now additionally receive the following rule:
+```
+apiGroups:
+- oauth.openshift.io
+resources:
+- useroauthaccesstokens
+verbs:
+- get
+- list
+- watch
+- delete
+```
+
+Since the new API allows to remove user-owned access tokens, it should no longer
+be necessary to keep the `system:oauth-token-deleter` clusterrole bound to the
+`system:authenticated` group. On the contrary, this opens a small security hole
```

Comment:
"this opens" sounds strange. Better: On the contrary, the currently existing cluster role binding means a remaining, small security hole where a user can delete tokens that he only knows the sha256 hash of (but it cannot be used to login). To address this security hole, ....